File: solutions/162 - Find Peak Element/Binary Search.cpp
#include <iostream>
#include <vector>
#include <sstream>
using namespace std;

class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        // Time Complexity: O(LogN) since we can safely halve the search space every iteration.
        // Space Complexity: O(1) for a few variables.
        if (nums.size() == 1) {
            return 0;
        }

        int l = 0;
        int r = nums.size() - 1;
        while (l <= r) {
            int m = l + (r - l) / 2;
            if (m - 1 >= 0 && m + 1 <= nums.size() - 1 && nums[m] > nums[m - 1] && nums[m] > nums[m + 1]) {
                return m;
            }
            if (m == 0 && m + 1 <= nums.size() - 1 && nums[m] > nums[m + 1]) {
                return m;
            }
            if (m == nums.size() - 1 && m - 1 >= 0 && nums[m] > nums[m - 1]) {
                return m;
            }
            if (m - 1 >= 0 && nums[m - 1] > nums[m]) {
                r = m - 1;
            } else if (m + 1 <= nums.size() - 1 && nums[m + 1] > nums[m]) {
                l = m + 1;
            }
        }
        return l;
    }
};

int main() {
    cout << "Enter nums: " << endl;
    string line;
    getline(cin, line);

    stringstream ss(line);
    vector<int> nums;
    int x;

    while (ss >> x) {
        nums.push_back(x);
    }

    Solution sol;

    cout << "Peak index: " << endl;
    cout << sol.findPeakElement(nums) << endl;

    return 0;
}

File: solutions/162 - Find Peak Element/Binary Search.java
import java.util.*;
class Solution {
    public int findPeakElement(int[] nums) {
        // Time Complexity: O(LogN) because we can use binary search.
        // Space Complexity: O(1) for a few variables.
        int left = 0;
        int right = nums.length - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] < nums[mid + 1]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter integers: ");
        String input = sc.nextLine();

        String[] parts = input.trim().split("\\s+");
        int[] nums = new int[parts.length];

        for (int i = 0; i < parts.length; i++) {
            nums[i] = Integer.parseInt(parts[i]);
        }

        Solution sol = new Solution();
        int peakIndex = sol.findPeakElement(nums);

        System.out.println("Peak element index: " + peakIndex);
        System.out.println("Peak element value: " + nums[peakIndex]);

        sc.close();
    }
}